src/main/java/dev/morling/onebrc/CalculateAverage_royvanrijn.java

```diff
@@ -50,30 +63,99 @@ private double round(double value) {
         }
     }
 
-    public static void main(String[] args) throws IOException {
+    public static void main(String[] args) throws Exception {
+
+        long before = System.currentTimeMillis();
+
+        // Split the file into numberOfChunks, use random access to read and process in parallel.
+
+        int numberOfChunks = 42; // Of course.
+
+        long fileSize = new File(FILE).length();
+        long chunkSize = fileSize / numberOfChunks;
+
+        List<Future<Map<String, Measurement>>> futures = new ArrayList<>();
 
-//        long before = System.currentTimeMillis();
+        ExecutorService executor = Executors.newVirtualThreadPerTaskExecutor();
 
-        Map<String, Measurement> resultMap = Files.lines(Path.of(FILE)).parallel()
-                .map(record -> {
-                    // Map to <String,double>
-                    int pivot = record.indexOf(";");
-                    String key = record.substring(0, pivot);
-                    double measured = Double.parseDouble(record.substring(pivot + 1));
-                    return new AbstractMap.SimpleEntry<>(key, measured);
-                })
-                .collect(Collectors.toConcurrentMap(
-                        // Combine/reduce:
-                        AbstractMap.SimpleEntry::getKey,
-                        entry -> new Measurement(entry.getValue()),
-                        Measurement::combineWith));
+        long lastEnd = 0;
+        for (int i = 0; i < numberOfChunks; i++) {
+            // Method findNextLineStart finds the correct point to break;
+            long start = lastEnd;
+            long end = (i == numberOfChunks - 1) ? fileSize : findNextLineStart(FILE, (i + 1) * chunkSize);
+            futures.add(executor.submit(() -> processChunk(FILE, start, end)));
+            lastEnd = end;
+        }
+
+        // Combine results from all futures, is there a better way to do this?
+        Map<String, Measurement> combinedResult = new TreeMap<>();
+        for (Future<Map<String, Measurement>> future : futures) {
+            final Map<String, Measurement> resultMap = future.get();
+            for (Map.Entry<String, Measurement> entry : resultMap.entrySet()) {
+                combinedResult.merge(entry.getKey(), entry.getValue(), Measurement::combineWith);
+            }
+        }
```

Comment:
> // Combine results from all futures, is there a better way to do this?

Maybe a bit more elegant using ExecutorCompletionService

```
var cs = new ExecutorCompletionService<Map<String, Measurement>>(executor);

long lastEnd = 0;
for (int i = 0; i < numberOfChunks; i++) {
	// Method findNextLineStart finds the correct point to break;
	long start = lastEnd;
	long end = (i == numberOfChunks - 1) ? fileSize : findNextLineStart(FILE, (i + 1) * chunkSize);
	cs.submit(() -> processChunk(FILE, start, end));
	lastEnd = end;
}

for (int i = 0; i < numberOfChunks; i++) {
	for (var entry : cs.take().get().entrySet()) {
		combinedResult.merge(entry.getKey(), entry.getValue(), Measurement::combineWith);
	}
}
```

Reply:
Ah yes, that is better, doesn't improve performance, but feels better :)